src/gui/TextEditor.cpp

```diff
@@ -960,53 +966,51 @@ void TextEditor::paint(cairo_t* cr, GdkRectangle* repaintRect, double zoom) {
             pos += gtk_text_iter_get_bytes_in_line(&cursorIter);
         }
         gtk_text_iter_set_offset(&cursorIter, offset);
-
         string txt = text.substr(0, pos) + preeditString + text.substr(pos);
 
-        PangoAttribute* attrib = pango_attr_underline_new(PANGO_UNDERLINE_SINGLE);
-        PangoAttrList* list = pango_layout_get_attributes(this->layout);
-
-        attrib->start_index = pos;
-        attrib->end_index = pos + preeditString.length();
-
-        if (list == nullptr) {
-            list = pango_attr_list_new();
-            pango_layout_set_attributes(this->layout, list);
-        }
-        pango_attr_list_insert(list, attrib);
-
+        PangoAttrList* attrlist = pango_attr_list_new();
+        PangoAttrList* preedit_attrlist = this->preeditAttrList;
+        pango_attr_list_splice(attrlist, preedit_attrlist, pos, preeditString.length());
+        pango_layout_set_attributes(this->layout, attrlist);
+        pango_attr_list_unref(attrlist);
+        attrlist = nullptr;
         pango_layout_set_text(this->layout, txt.c_str(), txt.length());
     } else {
         string txt = this->text->getText();
         pango_layout_set_text(this->layout, txt.c_str(), txt.length());
-    }
 
-    GtkTextIter start;
-    GtkTextIter end;
-    bool hasSelection = gtk_text_buffer_get_selection_bounds(this->buffer, &start, &end);
+        GtkTextIter start;
+        GtkTextIter end;
+        bool hasSelection = gtk_text_buffer_get_selection_bounds(this->buffer, &start, &end);
 
-    if (hasSelection) {
-        auto selectionColorU16 = Util::GdkRGBA_to_ColorU16(selectionColor);
-        PangoAttribute* attrib =
-                pango_attr_background_new(selectionColorU16.red, selectionColorU16.green, selectionColorU16.blue);
-        PangoAttrList* list = pango_layout_get_attributes(this->layout);
+        if (hasSelection) {
+            auto selectionColorU16 = Util::GdkRGBA_to_ColorU16(selectionColor);
+            PangoAttribute* attrib =
+                    pango_attr_background_new(selectionColorU16.red, selectionColorU16.green, selectionColorU16.blue);
+            PangoAttrList* attrlist = pango_layout_get_attributes(this->layout);
 
-        attrib->start_index = getByteOffset(gtk_text_iter_get_offset(&start));
-        attrib->end_index = getByteOffset(gtk_text_iter_get_offset(&end));
+            attrib->start_index = getByteOffset(gtk_text_iter_get_offset(&start));
+            attrib->end_index = getByteOffset(gtk_text_iter_get_offset(&end));
 
-        if (list == nullptr) {
-            list = pango_attr_list_new();
-            pango_layout_set_attributes(this->layout, list);
+            if (attrlist == nullptr) {
+                attrlist = pango_attr_list_new();
+                pango_layout_set_attributes(this->layout, attrlist);
+                pango_attr_list_unref(attrlist);
+                attrlist = nullptr;
+            }
+            pango_attr_list_insert(attrlist, attrib);
```

Comment:
This doesn't seem quite right. If `attrlist` is null above, it will be set to null again.
```suggestion
            const bool isNewAttrlist = attrlist == nullptr;
            if (isNewAttrlist) {
                attrlist = pango_attr_list_new();
                pango_layout_set_attributes(this->layout, attrlist);
            }
            pango_attr_list_insert(attrlist, attrib);
            if (isNewAttrlist) {
                pango_attr_list_unref(attrlist);
                attrlist = nullptr;
            }
```

Reply:
How fool I am! Thanks.